src/java/org/apache/cassandra/index/sai/disk/vector/CompressedVectorRowIdIterator.java

```diff
@@ -0,0 +1,145 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.disk.vector;
+
+import java.util.NoSuchElementException;
+import java.util.PriorityQueue;
+import java.util.function.IntFunction;
+
+import io.github.jbellis.jvector.vector.VectorSimilarityFunction;
+import org.apache.cassandra.utils.CloseableIterator;
+
+/**
+ * An iterator over {@link ScoredRowId} that lazily orders a {@link PriorityQueue} of approximately scored rows.
+ * The class works by:
+ * 1. Taking a priority queue of {@link ApproximateScore} ordered by approximate similarity score.
+ * 2. Polling the topK results from the approximate score queue
+ * 3. Materialize and score the full resolution vector for each row from step 2.
+ * 4. Add to the full resolution row {@link ScoredRowId} to the priority queue.
+ * 5. Return rows from the {@link ScoredRowId} priority queue until the limit is reached.
+ * 6. When the limit is reached, if the {@link ApproximateScore} queue is not empty, get the next (topK - limit) rows
+ *    from the approximate score queue and repeat 3-6 until the {@link ApproximateScore} queue is empty.
+ * 7. Return the remaining rows from the {@link ScoredRowId} queue.
+ */
+public class CompressedVectorRowIdIterator implements CloseableIterator<ScoredRowId>
+{
+    public static class ApproximateScore implements Comparable<ApproximateScore>
+    {
+        private final int rowId;
+        private final int ordinal;
+        private final float appoximateScore;
+
+        public ApproximateScore(int rowId, int ordinal, float appoximateScore)
+        {
+            this.rowId = rowId;
+            this.ordinal = ordinal;
+            this.appoximateScore = appoximateScore;
+        }
+
+        @Override
+        public int compareTo(ApproximateScore o)
+        {
+            // We always want scores in descending order
+            return Float.compare(o.appoximateScore, appoximateScore);
+        }
+    }
+
+    private final float[] queryVector;
+    // We use two PriorityQueues because we do not need an eager ordering of these results. Depending on how many
+    // sstables the query hits and the relative scores of vectors from those sstables, we may not need to return
+    // more than the first handful of scores.
+    // Priority queue with compressed vector scores
+    private final PriorityQueue<ApproximateScore> cvScoreQueue;
```

Comment:
suggest `approximateScoreQueue` and `exactScoreQueue`

Reply:
Good call, I'll update it